### PR TITLE
adds a dev mode collate for diagnostic info

### PR DESCRIPTION
**Pull request by wyli**
Signed-off-by: Wenqi Li <wenqil@nvidia.com>

### Description
follow-up of https://github.com/Project-MONAI/MONAI/issues/1833

this is a quick implementation to provide more info when collating fails.
if it's useful I can finish this PR with more tests and options

 
```py
import numpy as np
import torch
from monai.data.utils import dev_collate

dev_collate(
    [
        {"img": 2, "meta": {"shape": [torch.tensor(1.0)]}},
        {"img": 3, "meta": {"shape": [np.asarray(1.0)]}},
        {"img": 4, "meta": {"shape": [torch.tensor(1.0)]}},
    ]
)

```

logging diagnostic info:
```
> collate dict key "img" out of 2 keys
> collate dict key "meta" out of 2 keys
>> collate dict key "shape" out of 1 keys
>>> collate list of sizes: [1, 1, 1].
>>>> collate/stack a list of tensors
>>>> E: expected Tensor as element 1 in argument 0, but got numpy.ndarray, while stacking ['Tensor', 'ndarray', 'Tensor'] in collate((tensor(1.), array(1.), tensor(1.)))
```

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by wyli** [comment]
might be useful when developing https://github.com/Project-MONAI/MONAI/milestone/31

**Comment by ericspod** [comment]
I think this would be useful for sure.

**Review by ericspod — APPROVED** [review]

File: monai/data/utils.py [hunk 1]
```diff
@@ -254,6 +255,55 @@ def get_valid_patch_size(image_size: Sequence[int], patch_size: Union[Sequence[i
     return tuple(min(ms, ps or ms) for ms, ps in zip(image_size, patch_size_))
 
 
+def dev_collate(batch, level=1):
+    """collate with detailed error messages for debugging purposes."""
```
Comment (Nic-Ma) [inline monai/data/utils.py:259]: Please add doc-string for the `args` when you complete this PR.

Thanks.

**Review by Nic-Ma — APPROVED** [review]
Looks good to me except for a minor comment inline.

Thanks.

**Pull request merged**